### PR TITLE
Singlestore conn attrs

### PR DESCRIPTION
**Pull request by rohitsanj**
## Pull request checklist

Please check if your PR fulfills the following requirements:

- [x] Unit tests are present
- [x] Have you validated this change locally?

## Summary of Changes

- Create a new `SingleStoreDBConnection` class with preprocessing logic to add SingleStore connection attributes `program_name` and `program_version` corresponding to `noteable` and this package's version respectively.
- Replace `sqlalchemy-singlestoredb` installation with explicit `singlestoredb` installation and `sqlalchemy` as extras. 

Excerpt from an email from one of the SingleStore representatives:

>  We are requesting all SingleStore partners add a connection attribute to their connector to SingleStore, which will enable us to identify details about connections made to customers' SingleStore instances. It will help us track the health of the integration and joint customers. We will also be able to share these details across with Noteable.

### CODE REVIEW

**Comment by rohitsanj** [comment]
We need to bump `cryptography` to a minimum of 41.0.0 in order to avoid security vulnerabilities - https://pyup.io/v/59062/f17/. 

However, snowflake-connector-python upper bound restricts it from being updated to that version. There's an ongoing issue [here](https://github.com/snowflakedb/snowflake-connector-python/issues/1629) tracking a new release with upgraded cryptography version and one of the maintainers [mentioned](https://github.com/snowflakedb/snowflake-connector-python/issues/1629#issuecomment-1652170775) it should be out by Monday, if not earlier.

I guess.. we're blocked until then? @jlrobins @tizz98 

```
 name         : cryptography
 version      : 40.0.2
 description  : cryptography is a package which provides cryptographic recipes and primitives to Python developers.

dependencies
 - cffi >=1.12

required by
 - pyopenssl >=38.0.0,<40.0.0 || >40.0.0,<40.0.1 || >40.0.1,<42
 - snowflake-connector-python >=3.1.0,<41.0.0
```

**Comment by tizz98** [comment]
@rohitsanj yeah pretty blocked it seems, let's see again on monday

**Comment by rohitsanj** [comment]
Updated snowflake-connector-python to 3.1.0 and cryptography to 41.x

**Review by jlrobins — APPROVED** [review]

**Pull request merged**
